File: src/com/fate/engine/graphics/TextureResource.java
package com.fate.engine.graphics;

import static org.lwjgl.opengl.GL20.glDeleteProgram;

public class TextureResource {
	private int ID;
	private int count;
	
	public TextureResource(int ID) {
		this.ID = ID;
	}
	
	@Override
	public void finalize() {
		glDeleteProgram(ID);
	}
	
	public int getID() {
		return ID;
	}
	
	public void addReference() {
		count++;
	}
	
	public boolean removeReference() {
		count--;
		return count == 0;
	}
}

File: src/main/kotlin/com/morph/engine/graphics/TextureResource.kt
package com.morph.engine.graphics

import org.lwjgl.opengl.GL20.glDeleteProgram

class TextureResource(val id: Int) {
    private var count: Int = 0

    fun finalize() {
        glDeleteProgram(id)
    }

    fun addReference() {
        count++
    }

    fun removeReference(): Boolean {
        count--
        return count == 0
    }
}
